### ISSUE TITLE
Got error while go build about resolver/dns

### ISSUE THREAD

**Issue by EgorkZe** [grpc/grpc-go#3157]
### What version of gRPC are you using?
v1.25.0
### What version of Go are you using (`go version`)?
go1.13.3 darwin/amd64
### What operating system (Linux, Windows, …) and version?
Mac OS 
### What did you do?
If possible, provide a recipe for reproducing the error.
make go build
### What did you expect to see?
I expect to see build successful 
### What did you see instead?
 cannot load google.golang.org/grpc/resolver/dns: module google.golang.org/grpc@latest found (v1.25.0), but does not contain package google.golang.org/grpc/resolver/dns

**Comment by menghanl** [comment]
`google.golang.org/grpc/resolver/dns` was moved.

Do you reference this package in your code? If so, how are you using it? (If you need the builder, you can do a `resolver.Get("dns")`).

**Comment by owend** [comment]
This also breaks any packages that may still require an older version of grpc because they haven't updated yet and this change was only just released a couple of days ago.

grpc 1.anything still requires resolver/dns and not resolver/internal/dns so any build that requires it will now fail as well. 

Is there a way to leave old versions of those packages in place?  Right now it's a just a 404.

**Comment by dfawley** [comment]
We were never expecting any users to reference these packages directly.  We can re-add them in the old locations, but it would be good to understand why they are being referenced.

**Comment by owend** [comment]
I took a deeper look into this problem just because I was curious about the root cause and the error message here is a very weird one and made me doubt my mental model of how "dep" works with vendored dependencies.  I think my particular problem isn't the OP's problem, and mine is related to a wonky CI pipeline and not the grpc source code.   I thought it might have been related to using Gopkg rather than go modules but that's not it either.
  
The error happens ONLY inside an automated pipeline running in gitlab's environment.  I was not able to reproduce this error on my workstation using the same container image used in the CI pipeline.  Although this particular CI job doesn't use a cache I suspect that this is some wonkiness in the gitlab environment.  Anyway, whatever they do it can sometimes cause errors like this, probably from a mistakenly cached upstream package or URL.  I can just imagine the layers of CDN / reverse proxies / local mirrors etc that are going on behind the scenes there. 

TLDR - doesn't seem to be reproducible.

**Comment by owend** [comment]
Hooray it's non-deterministic!

"dep ensure -v" logs on working attempt it finds etcd package first and successfully picks a compatible grpc version 1.24

(16)  ? attempt go.etcd.io/etcd with 6 pkgs; 189 versions to try
(17)      try go.etcd.io/etcd@v3.4.3
(17)  ✓ select go.etcd.io/etcd@v3.4.3 w/83 pkgs
.....
(36)  ? attempt google.golang.org/grpc with 4 pkgs; 85 versions to try
(38)      try google.golang.org/grpc@v1.25.0
(39)  ✗   google.golang.org/grpc at v1.25.0 has problem subpkg(s):
(39)    	google.golang.org/grpc/resolver/dns is missing; required by go.etcd.io/etcd@v3.4.3.	google.golang.org/grpc/resolver/passthrough is missing; required by go.etcd.io/etcd@v3.4.3.
(38)      try google.golang.org/grpc@v1.24.0
(38)  ✓ select google.golang.org/grpc@v1.24.0 w/33 pkgs
....
  ✓ found solution with 276 packages from 64 projects

"dep ensure -v" logs on non-working attempt it finds grpc first, locks on 1.25 and then attempts to resolve etc which does not work with that version.  Not sure why it doesn't walk all the way back up the tree at this point but.... 🤷‍♂ 

(12)  ? attempt google.golang.org/grpc with 4 pkgs; 85 versions to try
(12)      try google.golang.org/grpc@v1.25.0
(12)  ✓ select google.golang.org/grpc@v1.25.0 w/35 pkgs
....
(19)  ? attempt go.etcd.io/etcd with 6 pkgs; 189 versions to try
(21)      try go.etcd.io/etcd@v3.4.3
(22)  ✗   go.etcd.io/etcd@v3.4.3 depping on google.golang.org/grpc at v1.25.0 has problem subpkg(s):	google.golang.org/grpc/resolver/dns is missing	google.golang.org/grpc/resolver/passthrough is missing
(21)      try go.etcd.io/etcd@v3.4.2
(22)  ✗   go.etcd.io/etcd@v3.4.2 depping on google.golang.org/grpc at v1.25.0 has problem subpkg(s):	google.golang.org/grpc/resolver/dns is missing	google.golang.org/grpc/resolver/passthrough is missing
(21)      try go.etcd.io/etcd@v3.4.1
(22)  ✗   go.etcd.io/etcd@v3.4.1 depping on google.golang.org/grpc at v1.25.0 has problem subpkg(s):	google.golang.org/grpc/resolver/dns is missing	google.golang.org/grpc/resolver/passthrough is missing
(21)      try go.etcd.io/etcd@v3.4.0
(22)  ✗   go.etcd.io/etcd@v3.4.0 not allowed by constraint ^3.4.1:
(22)      ^3.4.1 from (root)
....  at this point it tries a hundred old versions but none work
(21)    ← no more versions of go.etcd.io/etcd to try; begin backtrack
  ✗ solving failed
Solving failure: No versions of go.etcd.io/etcd met constraints:

**Comment by owend** [comment]
The mirror issue in etcd is here: https://github.com/etcd-io/etcd/issues/11336

Because it's an internal package I think the fix should be on the etcd side of this.  I don't have time this week but if someone does put together a patch, I wanted to link that issue here.

**Comment by menghanl** [comment]
The moved packages will be added back in #3162, to avoid the caused breakages. It would still be preferred to use only the public APIs, but that will become a minor issue.

**Comment by EgorkZe** [comment]
And so, what I should do to fix this problem in my code?

**Comment by dfawley** [comment]
Updating to 1.25.1 (after it is released) _should_ fix everything.

**Comment by menghanl** [comment]
And it this build failure is bothering you right now, pin gRPC to 1.24, until 1.25.1 is released.

**Comment by menghanl** [comment]
Reopen until 1.25.1 is released

**Comment by menghanl** [comment]
https://github.com/grpc/grpc-go/releases/tag/v1.25.1 is released. Please try again.
Closing issue now.
